Userland/Libraries/LibCore/EventLoopImplementationUnix.cpp

```diff
@@ -505,12 +514,34 @@ bool EventLoopManagerUnix::unregister_timer(int timer_id)
 
 void EventLoopManagerUnix::register_notifier(Notifier& notifier)
 {
-    ThreadData::the().notifiers.set(&notifier);
+    auto& thread_data = ThreadData::the();
+
+    thread_data.notifier_by_ptr.set(&notifier, thread_data.poll_fds.size());
+    thread_data.notifier_by_index.append(&notifier);
+    thread_data.poll_fds.append({
+        .fd = notifier.fd(),
+        .events = notification_type_to_poll_events(notifier.type()),
+        .revents = 0,
+    });
 }
 
 void EventLoopManagerUnix::unregister_notifier(Notifier& notifier)
 {
-    ThreadData::the().notifiers.remove(&notifier);
+    auto& thread_data = ThreadData::the();
+
+    auto it = thread_data.notifier_by_ptr.find(&notifier);
+    VERIFY(it != thread_data.notifier_by_ptr.end());
+
+    size_t notifier_index = it->value;
+    thread_data.notifier_by_ptr.remove(it);
+
+    if (notifier_index + 1 != thread_data.poll_fds.size()) {
+        swap(thread_data.poll_fds[notifier_index], thread_data.poll_fds.last());
+        swap(thread_data.notifier_by_index[notifier_index], thread_data.notifier_by_index.last());
+        thread_data.notifier_by_ptr.set(thread_data.notifier_by_index[notifier_index], notifier_index);
+    }
+    thread_data.poll_fds.take_last();
+    thread_data.notifier_by_index.take_last();
```

Comment:
Here we're modifying the pollfds structure that gets passed to the kernel right?

my 'possible race condition' alarm is going off... but I don't think it's possible for another thread to mess with this list right? ... I doubt we properly handle notifiers and multiple threads remotely correctly anyway.

Mostly looking for a confirmation that we can't possibly call this while the poll() is ongoing.

Perhaps a warning comment that we are modifying the actual list we will pass to poll is in order.

Reply:
> Mostly looking for a confirmation that we can't possibly call this while the poll() is ongoing.

I doubt this will be the first data race if someone tries to modify EventLoop concurrently :)

Reply:
Oh, and the structure being modified is literally called ThreadData, so I think comments will be excessive.